source/fuzz/transformation_wrap_vector_synonym.cpp

```diff
@@ -0,0 +1,151 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_wrap_vector_synonym.h"
+
+#include "source/fuzz/data_descriptor.h"
+#include "source/fuzz/fuzzer_util.h"
+#include "source/opt/instruction.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationWrapVectorSynonym::TransformationWrapVectorSynonym(
+    protobufs::TransformationWrapVectorSynonym message)
+    : message_(std::move(message)) {}
+
+TransformationWrapVectorSynonym::TransformationWrapVectorSynonym(
+    uint32_t instruction_id, uint32_t vector_operand1, uint32_t vector_operand2,
+    uint32_t fresh_id, uint32_t pos) {
+  message_.set_instruction_id(instruction_id);
+  message_.set_vector_operand1(vector_operand1);
+  message_.set_vector_operand2(vector_operand2);
+  message_.set_fresh_id(fresh_id);
+  message_.set_scalar_position(pos);
+}
+
+bool TransformationWrapVectorSynonym::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  auto instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.instruction_id());
+  auto vec1 = ir_context->get_def_use_mgr()->GetDef(message_.vector_operand1());
+  auto vec2 = ir_context->get_def_use_mgr()->GetDef(message_.vector_operand2());
+
+  // |instruction_id| must refer to an existing instruction.
+  if (instruction == nullptr) {
+    return false;
+  }
+
+  auto type_instruction =
+      ir_context->get_def_use_mgr()->GetDef(instruction->type_id());
+
+  // The instruction must be of a valid scalar operation type.
+  if (!OpcodeIsSupported(instruction->opcode()) ||
+      !OperandTypeIsSupported(type_instruction)) {
+    return false;
+  }
+
+  assert(!transformation_context.GetFactManager()->IdIsIrrelevant(
+             instruction->result_id()) &&
+         "Result id of the scalar operation must be relevant.");
+
+  // |vector_operand1| and |vector_operand2| must exist.
+  if (vec1 == nullptr || vec2 == nullptr) {
+    return false;
+  }
+
+  // The 2 vectors must be the same valid vector type.
+  auto vec1_type_id = vec1->type_id();
+  auto vec2_type_id = vec2->type_id();
+
+  if (vec1_type_id != vec2_type_id) {
+    return false;
+  }
+
+  auto vec_type = ir_context->get_def_use_mgr()->GetDef(vec1_type_id)->opcode();
+
+  if (vec_type != SpvOpTypeVector) {
+    return false;
+  }
+
+  // |fresh_id| must be fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.fresh_id())) {
+    return false;
+  }
+
+  // |scalar_position| needs to be a non-negative integer less than the vector
+  // length.
+  // OpTypeVector instruction has the component count at index 2.
+  auto vec_len = ir_context->get_def_use_mgr()
+                     ->GetDef(vec1_type_id)
+                     ->GetSingleWordInOperand(0);
+  if (message_.scalar_position() >= vec_len) {
```

Comment:
Similarly, I would probably get rid of the `vec_len` temporary and use the expression inline in the `if` statement, but this is a matter of taste, so happy with whatever you decide.

Reply:
I think inline is indeed better. The temporary is only being used once and the comment explains pretty clearly that it is used to check in case ``|scalar_position|`` go out of bounds.